File: GameServer/pyserver/lib/baseInstance.py
import config
from gameState import GameState


class BaseInstance():
    # array of user shared keys
    users = []
    status = ''

    def __init__(self, session_key, game_state_params=config.GAME_STATE_PARAMS):
        self.session_key = session_key
        self.game_state = GameState(game_state_params)

    # add user to instance
    def join_instance(self, client_shared_key):
        self.users.append(client_shared_key)

    def leave_instance(self, client_shared_key):
        self.users.remove(client_shared_key)
        # todo test this
        # todo remove all the removed instances
        if len(self.users) == 0:
            self.status = 'removed'

File: GameServer/cpp/baseInstance.cpp

using namespace std;

class BaseInstance{
	public:
	vector<string> users;
	string status;
	string session_key;
	GameState game_state;

	BaseInstance(string);
	void join_instance(string);
	void leave_instance(string);
};

BaseInstance::BaseInstance(string server_session_key){
	session_key = server_session_key;
	game_state = GameState();
}

void BaseInstance::join_instance(string client_shared_key){
	status = "";
	users.push_back(client_shared_key);
}
void BaseInstance::leave_instance(string client_shared_key){
	for(vector<string>::iterator search = users.begin(); search != users.end(); ++search){
		if(*search == client_shared_key){
			users.erase(search);	
			break;
		}
	}
	if(users.begin() == users.end())
		status = "removed";
}
